### DISCUSSION TITLE
how to determine the date and time when the moon will be in a certain part of the sky

### DISCUSSION THREAD

**Discussion by David-Valters** [cosinekitty/astronomy#308] [Q&A]
how to determine the date and time when the moon will appear and leave a certain area of the sky, which is behind two elevation (e1, e2) and two azimuths (a1, a2), you need to calculate all events within 27 days, there may also be a situation that the moon will never appear in the given area.
I will gladly buy coffee for your help.
a - azimuth
e - elevation
![photo_2023-06-17_19-23-54](https://github.com/cosinekitty/astronomy/assets/46930139/ffd9b366-b678-41be-b9c9-d76fc4b64c9d)

**Comment by cosinekitty** [comment]
This is an interesting problem! I will think about the best way to solve it...

**Comment by cosinekitty** [comment]
One question: which programming language do you want to use?

**Reply by David-Valters** [reply]
I plan to use C++, I apologize for the impudence, it would be great if you could tell me how to solve this problem by Wednesday, I would use your library and mention it in the list of technologies used in my graduate work

**Reply by cosinekitty** [reply]
I have thought about this quite a bit. Because your development time is limited, and there is a complicated mix of special cases, the safest approach (meaning, the most likely to be correct) will be a "brute force" solution. With this in mind, I have a few questions for you:

* If the center of the Moon is within your alt/az window for a very small amount of time, is it tolerable for you to ignore the event? If so, what would the tolerance be: one minute, one second, one millisecond, ...? The smaller the tolerance is, the less efficient the search will be.
* Do you ever have cases where the azimuth range includes due north? This can make the algorithm more complicated because the azimuth can increase up to 359.9999...&deg;, but then drop back to 0&deg;.

I'm thinking to start with a very simple but inefficient algorithm. Create a boolean function that tells you whether the Moon is inside the window or not. You give it an observer and a time, and it returns true/false. You could even make this a member function of a C++ class that contains the observer and alt/az window.

Then start at a given start time and call the function. Keep adding &Delta;t and calling the function until it returns true. Now you have a time bound (t, t+&Delta;t) that straddles the boundary between being outside the window and inside the window. Do a binary search inside this time range to find the transition time to whatever arbitrary precision you want (perhaps one millisecond).

Of course this will be rather slow, but once you get it debugged and working, you can greatly improve it by estimating how far away the Moon is from your window at any given time. If the Moon's altitude is 30 degrees away from the closest altitude limit (subtract and take absolute value), and 45 degrees away from the closest azimuth limit, then you know the Moon must move at least 30 degrees in the sky to reach the window.

The Moon is always in retrograde movement with respect to background stars, so it always takes more than 24 hours to pass through 360 degrees of sky, so 15 degrees per hour is a generous overestimate of its angular speed. Therefore, it can't possibly move 30 degrees in less than 2 hours. So you could safely step 2 hours in this case. This trick will greatly speed up your search.

I hope this helps. I know it's only a start, but I can answer follow up questions as needed.

**Reply by cosinekitty** [reply]
As a refinement, it would be possible to search for the next time when the Moon is closest to the center of your window. Then you could call the function to see whether it is actually inside the window at that time. This would be much faster.

**Reply by David-Valters** [reply]
> If the center of the Moon is within your alt/az window for a very small amount of time, is it tolerable for you to ignore the event? If so, what would the tolerance be: one minute, one second, one millisecond, ...? The smaller the tolerance is, the less efficient the search will be.

an event that will last less than 3-5 minutes can be ignored
>Do you ever have cases where the azimuth range includes due north? This can make the algorithm more complicated because the azimuth can increase up to 359.9999...°, but then drop back to 0°.

I think north can be included, since the data will be entered by the user, but the values ​​can only be integers, in general, the program needs to search for no more than 4-7 seconds

**Comment by cosinekitty** [comment]
Try taking a look at [altazsearch.cpp](https://github.com/cosinekitty/astronomy/blob/master/demo/c/altazsearch.cpp), I demo program I just added. It implements the idea of doing a two-part search. In the first part, it finds when the body passes closest to the center of your alt/az window.

If the body is inside the window at that time, it steps forward and backward using a pair of binary searches to find when the body will enter and leave the window. It also handles the case where the azimuth range includes north. To represent this, specify `az1 > az2`.  For example, if az1=10, az2=350, then the azimuth range is very wide and wraps around almost the whole sky. But if az1=350 and az2=10, you get a 20&deg; wide azimuth range that includes due north.

I hope this helps you with your project!
